src/kernel-install/kernel-install.c

```diff
@@ -0,0 +1,1085 @@
+         * they point to the ESP. */
+        FOREACH_STRING(p, "/efi", "/boot/efi") {
+                r = path_is_mount_point(p, /* root = */ NULL, /* flags = */ 0);
+                if (r <= 0) {
+                        if (IN_SET(r, 0, -ENOENT, -ENOTDIR))
+                                continue;
+
+                        return log_error_errno(r, "Failed to check if '%s' is a mount point: %m", p);
+                }
+
+                c->boot_root = strdup(p);
+                if (!c->boot_root)
+                        return log_oom();
+
+                log_debug("%s is a mount point, using BOOT_ROOT=%s.", c->boot_root, c->boot_root);
+                return 0;
+        }
+
+        /* If all else fails, use /boot. */
+        c->boot_root = strdup("/boot");
+        if (!c->boot_root)
+                return log_oom();
+
+        log_debug("KERNEL_INSTALL_BOOT_ROOT autodetection yielded no candidates, using \"%s\".", c->boot_root);
+        return 0;
+}
+
+static int context_load_entry_token(Context *c) {
+        _cleanup_free_ char *path = NULL, *token = NULL;
+        int r;
+
+        assert(c);
+        assert(!c->entry_token);
+
+        path = path_join(c->conf_root ?: "/etc/kernel", "entry-token");
+        if (!path)
+                return log_oom();
+
+        r = read_one_line_file(path, &token);
+        if (r == -ENOENT)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to read %s: %m", path);
+
+        if (isempty(token))
+                return 0;
+
+        if (!string_is_safe(token))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "%s contains unsafe character(s).", path);
+
+        if (!filename_is_valid(token))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid entry token read from %s: %s", path, token);
+
+        log_debug("entry-token \"%s\" acquired from %s.", token, path);
+        c->entry_token = TAKE_PTR(token);
+        return 1; /* loaded */
+}
+
+static int context_ensure_entry_token(Context *c) {
+        assert(c);
+
+        if (c->entry_token)
+                return 0;
+
+        c->entry_token = strdup(SD_ID128_TO_STRING(c->machine_id));
+        if (!c->entry_token)
+                return log_oom();
+
+        log_debug("No entry-token candidate matched, using \"%s\" from machine-id.", c->entry_token);
+        return 0;
+}
+
+static int context_load_plugins(Context *c) {
+        int r;
+
+        assert(c);
+
+        if (c->plugins)
+                return 0;
+
+        r = conf_files_list_strv(&c->plugins,
+                                 ".install",
+                                 /* root = */ NULL,
+                                 CONF_FILES_EXECUTABLE | CONF_FILES_REGULAR | CONF_FILES_FILTER_MASKED,
+                                 STRV_MAKE_CONST("/usr/lib/kernel/install.d", "/etc/kernel/install.d"));
```

Comment:
maybe use `CONF_PATHS_STRV("kernel/install.d")`? so that people can drop in plugins into /run/ too, and to make this systematic?